Add support for sampling timers

a-kr opened the pull request:
**Problem**:

`gostatsd` does not support sample rate for timer metrics, only for counters.
Original Etsy's `statsd` implementation supports sample rate for timers since 2013: https://github.com/etsy/statsd/pull/243/commits/cc12534e3b3bf5137a263e4e065876329e1a719c

**Situation**:

Consider a generic API service which does performance tracking by sending a single `timer` metric for each request to its endpoint. Common statsd-carbon-graphite pipeline (or any equivalent one) produces both mean / percentile response time data and requests-per-second data from `timer` metrics, which is sufficient for most monitoring situations.

Now, suppose this service handles quite a lot of requests per second. In this situation is reasonable to use metric sampling and submit only, say, 50% of timer metrics:

    megaservice.supernode-14.get_customers:110|ms|@0.5
    
This works well with Etsy's `statsd`: values of `count` and `count_ps` generated from a timer metric are divided by the sample rate, producing a (reasonable estimate of a) correct requests-per-second data. `mean`, `median`, `upper_95`, `sum`, and other aggregated values which are measured in seconds are not affected by sampling.

With `gostatsd`, however, sampling works only for counters and not for timers. So if I use `gostatsd` and submit timer metrics with 50% sample rate, I get `count_ps` values which are 2 times lower than the actual number of requests per second.

**Note**:

Etsy's `statsd` implementation assumes that, within a single flush interval, a single timer metric is submitted with the same sample rate:

    megaservice.supernode-14.get_customers:54|ms|@0.5
    megaservice.supernode-14.get_customers:22|ms|@0.5
    megaservice.supernode-14.get_customers:5553|ms|@0.5

If this assumption holds, then only the counters (`count`, `count_ps`) need to be adjusted for samping, and mean / percentile / etc time values can be calculated as normal.

If this assumption breaks, and we recieve metrics with different sample rates within a single flush interval...

    megaservice.supernode-14.get_customers:54|ms|@0.1
    megaservice.supernode-14.get_customers:22|ms|@0.5
    megaservice.supernode-14.get_customers:5553|ms|@0.25

...then all aggregations (mean / percentile / etc) should be somehow computed with regard to the sample rate of each individual metric: mean becomes weighted average, etc; or else they would not make sense statistically.

In practice, it is hard to imagine why would one need to change the sample rate continuously, and how would that be implemented on the client side. Most likely the sample rate is chosen by a developer and stays the same for long periods of time. When it does eventually change, only a single flush interval would be affected.

So I've chosen to keep the assumption of a single, rarely-changing sample rate per metric --- same assumption that is used in Etsy's `statsd`.

**Summary**:

This pull request implements support for sample rate for timer counters in `gostatsd`, in original-Etsy's-`statsd`-compatible way.

tiedotguy commented on pkg/statsd/aggregator.go:
```diff
@@ -284,20 +287,23 @@ func (a *MetricAggregator) receiveGauge(m *gostatsd.Metric, tagsKey string, now
-			t.Timestamp = now
-		} else {
-			t = gostatsd.NewTimer(now, []float64{m.Value}, m.Hostname, m.Tags)
+	if !ok {
```
This restructuring changes it away from the other 3 ```receive*``` calls.  I generally agree that it's a cleaner pattern, however historically whenever I've done this, there's a performance impact (not a lot, but present) that I can't justify just for the style change.

a-kr replied:
Fixed, reverted `receiveTimer` to original structure

tiedotguy commented:
Can you please also ensure the CLA is signed as appropriate.

a-kr commented:
Thanks for the review.
* Added a note to the changelog;
* reverted refactoring in `receiveTimer`;
* pushed the changes as `fixup!` commits (will squash them if everything is ok).

Signed _Atlassian Individual CLA_.

tiedotguy commented:
I'll keep an eye out for the CLA on our side (unfortunately it's still a bit manual), but it looks good to me. 
 I'll get it released once that shows up.

Thanks!

a-kr commented:
Any news on CLA? If it helps, I've signed it on August 28, as Alexey Kryuchkov.

Not sure what can be done about the failing CI build --- `make test-race` consistently passes on my machine. Any way to restart it?

tiedotguy commented:
Sorry for the delays - finally came in overnight.  Not sure about the test, looks like a flake.  Trying to repro it locally.  If I can I'll raise an issue, otherwise call it cosmic rays from space.
